lib/sinatra/base.rb

```diff
@@ -1019,7 +1019,13 @@ def route_eval
     def process_route(pattern, conditions, block = nil, values = [])
       route = @request.path_info
       route = '/' if route.empty? and not settings.empty_path_info?
-      route = route[0..-2] if !settings.strict_paths? && route != '/' && route.end_with?('/')
+      if !settings.strict_paths? && route != '/'
+        if pattern.to_s.end_with?('/') && !route.end_with?('/')
+          route << '/'
+        elsif !pattern.to_s.end_with?('/') && route.end_with?('/')
+           route = route[0..-2]
+        end
+      end
```

Comment:
Please reduce same processings by reflecting the following patch:

```diff
diff --git lib/sinatra/base.rb lib/sinatra/base.rb
index 3bcb7e99..c314aad1 100644
--- lib/sinatra/base.rb
+++ lib/sinatra/base.rb
@@ -1020,10 +1020,12 @@ module Sinatra
       route = @request.path_info
       route = '/' if route.empty? and not settings.empty_path_info?
       if !settings.strict_paths? && route != '/'
-        if pattern.to_s.end_with?('/') && !route.end_with?('/')
+        pattern_end_with_slash = pattern.to_s.end_with?('/')
+        route_end_with_slash = route.end_with?('/')
+        if pattern_end_with_slash && !route_end_with_slash
           route << '/'
-        elsif !pattern.to_s.end_with?('/') && route.end_with?('/')
-           route = route[0..-2]
+        elsif !pattern_end_with_slash && route_end_with_slash
+          route = route[0..-2]
         end
       end
       return unless params = pattern.params(route)
```

However, I'm not sure if the changed behavior would be good because the `pattern` will contain `Regexp` pattern not only the `String` pattern.